### ISSUE TITLE
Lazy query run multiple times on scrolling

### ISSUE THREAD

**Issue by elmarcel** [tlaukkan/vaadin-lazyquerycontainer#75]
Hello,
I'm developing a Grid to display, filter and select items from an Oracle DB table of 2,5 million records (growing).
The lazy query structure is:

```sql
select *
from
(
select rownum as rnum, id, field1, field2, etc
from table t
<dynamic where clause>
order by id
) where rnum >= ? and rnum <= ?
```

the two parameters are set with `startIndex+1`  and `startIndex+count` in my Query implementation.
My batch size is set to 500.
When I first access the page containing the Grid I found in my logs that the `loadItems` method is invoked multiple times, which is not something I expected.

```
2015-07-16 12:05:03,147 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=0, count=500 - start
2015-07-16 12:05:06,063 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=0, count=500 - end
2015-07-16 12:05:06,428 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=0, count=500 - end result processing

-> scrollar moved down

2015-07-16 12:05:20,935 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - start
2015-07-16 12:05:23,327 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end
2015-07-16 12:05:23,670 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end result processing
2015-07-16 12:05:24,049 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - start
2015-07-16 12:05:26,933 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end
2015-07-16 12:05:27,186 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end result processing
2015-07-16 12:05:27,535 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - start
2015-07-16 12:05:29,935 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end
2015-07-16 12:05:30,149 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end result processing
2015-07-16 12:05:30,543 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - start
2015-07-16 12:05:33,342 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end
2015-07-16 12:05:33,513 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end result processing
2015-07-16 12:05:33,952 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - start
2015-07-16 12:05:36,512 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end
2015-07-16 12:05:36,792 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=1500, count=500 - end result processing
2015-07-16 12:05:37,249 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - start
2015-07-16 12:05:39,213 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end
2015-07-16 12:05:39,638 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=2000, count=500 - end result processing
2015-07-16 12:05:40,561 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - start
2015-07-16 12:05:43,128 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end
2015-07-16 12:05:43,490 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end result processing
2015-07-16 12:05:43,840 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - start
2015-07-16 12:05:46,224 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end
2015-07-16 12:05:46,536 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end result processing
2015-07-16 12:05:46,952 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - start
2015-07-16 12:05:49,410 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end
2015-07-16 12:05:49,622 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end result processing
2015-07-16 12:05:50,114 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - start
2015-07-16 12:05:52,610 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end
2015-07-16 12:05:52,788 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end result processing
2015-07-16 12:05:53,178 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - start
2015-07-16 12:05:55,409 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end
2015-07-16 12:05:55,738 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=102500, count=500 - end result processing
2015-07-16 12:05:56,187 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - start
2015-07-16 12:05:58,718 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end
2015-07-16 12:05:58,946 [http-bio-8080-exec-3545] TRACE myapp.ui.QueryImp - loadItems: startIndex=103000, count=500 - end result processing
```

As visibile from the timestamps, the whole operation lasts too long to make the application usable, while the time needed for a single query is still acceptable.
Is it possibile to make the container less "sensitive" to the scrollbar and performn queries only when this stops?
Kind regards,
Marco

**Comment by tlaukkan** [comment]
Hi

I am afraid the scroll bar related loading functionality is related to grid
implementation and not governed by the container.

You could try to increase the cache size of lazy query container.

You could try using lazy entity container and filtering with indexed fields
like for example time window to get less results to grid. Usually a bit
smaller batch size is good like 50 to 100 rows. In the end with large
tables you tend to need to filter efficiently. User cannot efficiently
browse thousands of rows. Often having mandatory prefilled filters which
narrow the result set are useful as they will limit the initial grid
content.

Br,
Tommi

**Comment by elmarcel** [comment]
Hi Tommi,
thanks for your replay. I tried to set the batch size smaller (down to 100) and bigger (up to 10000).
When it is smaller, query time decreases, but more queries are performed. It felt worse than 1000.
Whent it is bigger, less queries are performed, but query time increases (a lot).
I gave up, and implemented a custom paged container, since I cound not find a similar one in vaadin addins.
It is read only, but it works quite well. I will release it as add-on by the end of the project I'm working on.
Ciao,
Marco
